In [ ]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from IPython.display import display

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Part 1: Load the dataset

In [ ]:
# Load the dataset (load remotely or locally)
iris = pd.read_csv('/content/drive/MyDrive/University/Academics/Fall 2024/CS 383/Projects/CS383_P5/iris.csv')
# Output the first 15 rows of the data
display(iris.head(15))
# Display a summary of the table information (number of datapoints, etc.)
display(iris.info())


,sepal_length,sepal_width,petal_length,petal_width,species
0,5.1,3.5,1.4,0.2,setosa
1,4.9,3.0,1.4,0.2,setosa
2,4.7,3.2,1.3,0.2,setosa
3,4.6,3.1,1.5,0.2,setosa
4,5.0,3.6,1.4,0.2,setosa
5,5.4,3.9,1.7,0.4,setosa
6,4.6,3.4,1.4,0.3,setosa
7,5.0,3.4,1.5,0.2,setosa
8,4.4,2.9,1.4,0.2,setosa
9,4.9,3.1,1.5,0.1,setosa


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


None

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

The dataset used is the Iris dataset, which contains measurements of iris flowers from three different species: Setosa, Versicolor, and Virginica.
The features in the dataset are:
- sepal_length: Length of the sepal in centimeters.
- sepal_width: Width of the sepal in centimeters.
- petal_length: Length of the petal in centimeters.
- petal_width: Width of the petal in centimeters.

The label is the species of the iris flower, which can be one of the following:
- Setosa mapped as 0
- Versicolor mapped as 1
- Virginica mapped as 2

# Part 2: Split the dataset into train and test

In [71]:
# Take the dataset and split it into our features (X) and label (y)
X = iris[['sepal_length', 'sepal_width', 'petal_length', 'petal_width']]
y = iris['species']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=50)


# Part 3: Logistic Regression

In [72]:
# i. Use sklearn to train a LogisticRegression model on the training set
logreg = LogisticRegression()
logreg.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_datapoint = pd.DataFrame([[5.1, 3.5, 1.4, 0.2]], columns=['sepal_length', 'sepal_width', 'petal_length', 'petal_width'])
sample_datapoint_proba = logreg.predict_proba(sample_datapoint)
print("Logistic Regression Prediction for Sample:", sample_datapoint_proba)

# iii. Report on the score for Logistic regression model, what does the score measure?
logreg_score = logreg.score(X_test, y_test)
display("Logistic Regression Score:", logreg_score)

# iv. Extract the coefficients and intercepts for the boundary line(s)
logreg_coefficients = logreg.coef_
logreg_intercepts = logreg.intercept_
print("Logistic Regression Coefficients:", logreg_coefficients)
print("Logistic Regression Intercepts:", logreg_intercepts)

Logistic Regression Prediction for Sample: [[9.78537255e-01 2.14627201e-02 2.50077542e-08]]


'Logistic Regression Score:'

1.0

Logistic Regression Coefficients: [[-0.42326446  0.97656528 -2.45647874 -1.04890609]
 [ 0.51368556 -0.21905272 -0.20248426 -0.86080569]
 [-0.0904211  -0.75751256  2.65896301  1.90971178]]
Logistic Regression Intercepts: [  9.49023001   1.88349415 -11.37372416]


# Part 4: Support Vector Machine

In [89]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svm = SVC(probability=True)  # enabling probability if needed
svm.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_datapoint = pd.DataFrame([[5.1, 3.5, 1.4, 0.2]], columns=['sepal_length', 'sepal_width', 'petal_length', 'petal_width'])
sample_datapoint_proba_svm = svm.predict_proba(sample_datapoint)
print("SVM Prediction for sample:", sample_datapoint_proba_svm)

# iii. Report on the score for the SVM, what does the score measure?
svm_score = svm.score(X_test, y_test)
print("SVM Score:", svm_score)

SVM Prediction for sample: [[0.97455916 0.01637318 0.00906766]]
SVM Score: 1.0


# Part 5: Neural Network

In [90]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp = MLPClassifier(max_iter=1000, hidden_layer_sizes=(100), activation='tanh')
mlp.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_datapoint = pd.DataFrame([[5.1, 3.5, 1.4, 0.2]], columns=['sepal_length', 'sepal_width', 'petal_length', 'petal_width'])
sample_datapoint_proba = mlp.predict_proba(sample_datapoint)
print("MLPClassifier Prediction for Sample: ", sample_datapoint_proba)

# iii. Report on the score for the Neural Network, what does the score measure?
score = mlp.score(X_test, y_test)
print("Score: ",score)

# iv: Experiment with different options for the neural network, report on your best configuration
# THe default MLPClassifier was getting a score of 0.97. Experimenting with the max_iter parameter = 5 is 0.4 accuracy, 50 is 0.733, 500 is score 0.866.0
# I also experimented with the activation function and tanh gave a higher accuracy of 0.933
# Changing the learning rate alpha had no apparant effect on the acuracy

MLPClassifier Prediction for Sample:  [[9.98051970e-01 1.94802952e-03 1.06395919e-10]]
Score:  0.9333333333333333


# Part 6: K-Nearest Neighbors

In [91]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
knn = KNeighborsClassifier()
knn.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_datapoint = pd.DataFrame([[5.1, 3.5, 1.4, 0.2]], columns=['sepal_length', 'sepal_width', 'petal_length', 'petal_width'])
sample_datapoint_proba = knn.predict_proba(sample_datapoint)
print("KNN Prediction for Sample: ",sample_datapoint_proba)

# iii. Report on the score for kNN, what does the score measure?
score = knn.score(X_test, y_test)
print("Score: ",score)

KNN Prediction for Sample:  [[1. 0. 0.]]
Score:  0.9333333333333333


# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

The results show that the Support Vector Machine (SVM) and Logistic Regression models performed the best on the dataset, both achieving a perfect score of 1.0 on the test set. This indicates that these models were able to perfectly classify the iris species in the test set, which is something I have not seen in real world.

The K-Nearest Neighbors (KNN) and Neural Network (MLPClassifier) models also performed well, with scores of 0.9333. This suggests that these models were also effective but not as perfect as SVM and Logistic Regression.

The high performance of SVM and Logistic Regression could be attributed to the linear separability of the Iris dataset, which these models handle well. The KNN and Neural Network models, while powerful, might not have been as well-tuned or might require more data to achieve similar performance.

One surprising finding to me was the perfect scores achieved by SVM and Logistic Regression, which is not always expected in real-world datasets. I have never seen a perfect score on a ML model.


---

